### PR TITLE
[BPK-4228] Add BPKOverlayView

### PR DESCRIPTION
**Pull request by georgegillams**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] `UNRELEASED.md`
+ [ ] `README.md`
+ [ ] Tests
+ [x] [Screenshotting code](https://github.com/Skyscanner/backpack-ios/blob/master/Example/Backpack%20Screenshot/Screenshots.swift)
+ [x] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/master/Backpack/Backpack.h)
+ [ ] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack/OverlayView/Classes/BPKOverlayView.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+/**
+ * The kind of overlay to apply to the view.
+ */
+@property(nonatomic, assign) IBInspectable BPKOverlayViewOverlayType overlayType;
+
+/**
+ * The view to display in front of the tint.
+ */
+@property(nonatomic, strong, readonly) UIView *foregroundView;
```
Comment (georgegillams) [inline Backpack/OverlayView/Classes/BPKOverlayView.h:45]: I'm thinking we should create these and expose them, but not allow consumers to set their own.

Comment (k0nserv) [reply Backpack/OverlayView/Classes/BPKOverlayView.h:45]: Yeah 👍 Maybe make this one lazy though. Just like `contentView` and `backgroundView` in `UITableViewCell`.

File: Backpack/OverlayView/Classes/BPKOverlayView.m [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    _overlayType = overlayType;
+
+    CGFloat newOpacity = 0;
+    if (overlayType == BPKOverlayViewOverlayTypeTint) {
+        newOpacity = 0.56;
+    }
+
+    [CATransaction begin];
+    [CATransaction setAnimationDuration:0.2];
```
Comment (k0nserv) [inline Backpack/OverlayView/Classes/BPKOverlayView.m:72]: Noice 👍 Do we not have a token for 0.2s?

File: Backpack/OverlayView/Classes/BPKOverlayView.m [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+- (void)setupWithCornerStyle:(BPKOverlayViewCornerStyle)cornerStyle {
+    self.tintLayer = [CALayer layer];
+    self.tintLayer.backgroundColor = BPKColor.skyGrayTint02.CGColor;
+    self.tintLayer.opacity = 0;
+    [self.layer addSublayer:self.tintLayer];
+
+    self.layer.masksToBounds = NO;
+    self.backgroundColor = BPKColor.backgroundTertiaryColor;
+    [[BPKShadow shadowSm] applyToLayer:self.layer];
```
Comment (k0nserv) [inline Backpack/OverlayView/Classes/BPKOverlayView.m:91]: I didn't know this was supposed to have a shadow

File: Backpack/OverlayView/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Backpack/OverlayView
+
+## Usage
+
+`BPKOverlayView` provides an easy way to place content either in-front or behind a tint layer. It ensures that, when placing text on a graphic, the text will be legible and accessible.
```
Comment (shaundon) [inline Backpack/OverlayView/README.md:5]:
```suggestion
`BPKOverlayView` provides an easy way to place content either in front or behind a tint layer. It ensures that when placing text on a graphic, the text will be legible and accessible.
```

**Review by shaundon — APPROVED** [review]
Have run locally and it works well, although somebody with more UIKit familiarity like @k0nserv should probably look it over before approving too

File: Backpack/OverlayView/Classes/BPKOverlayView.m [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+- (UIView *)backgroundView {
+    if (_backgroundView != nil) {
+        return _backgroundView;
+    }
+
+    self.tintLayer = [CALayer layer];
+    self.tintLayer.backgroundColor = BPKColor.skyBlueShade03.CGColor;
+    self.tintLayer.zPosition = 100;
```
Comment (k0nserv) [inline Backpack/OverlayView/Classes/BPKOverlayView.m:116]: This feels arbitrary and magical 🤔

File: Backpack/OverlayView/Classes/BPKOverlayView.m [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        [self.trailingAnchor constraintEqualToAnchor:_backgroundView.trailingAnchor],
+        [self.bottomAnchor constraintEqualToAnchor:_backgroundView.bottomAnchor]
+    ]];
+
+    return _backgroundView;
+}
+
+#pragma mark - setters
+
+- (void)setCornerStyle:(BPKOverlayViewCornerStyle)cornerStyle {
```
Comment (k0nserv) [inline Backpack/OverlayView/Classes/BPKOverlayView.m:138]: This and several other methods are missing `BPKAssertMainThread`

File: Backpack/OverlayView/Classes/BPKOverlayView.m [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+#import <Backpack/Duration.h>
+#import <Backpack/Radii.h>
+#import <Backpack/Shadow.h>
+#import <Backpack/Spacing.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BPKOverlayView ()
+
+@property(nonatomic, strong, nullable) CALayer *tintLayer;
```
Comment (k0nserv) [inline Backpack/OverlayView/Classes/BPKOverlayView.m:31]: This feels a bit confusing. `BPKOverlayView` now owns a layer with `strong` memory semantics that is in fact a sublayer of a subview of `BPKOverlayView`, would it make more sense to move this ownership and responsbility into a new class and make `backgroundView` an instance of it?

File: Backpack/OverlayView/Classes/BPKOverlayView.m [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+    CGFloat cornerRadius = 0;
+
+    switch (self.cornerStyle) {
+        case BPKOverlayViewCornerStyleSmall:
+            cornerRadius = BPKCornerRadiusMd;
+            break;
+        case BPKOverlayViewCornerStyleLarge:
+            cornerRadius = BPKCornerRadiusLg;
+            break;
+        default:
```
Comment (k0nserv) [inline Backpack/OverlayView/Classes/BPKOverlayView.m:184]: Maybe add an assert here to catch us if we forget to update this? I think the compiler might already help, but that's worth verifying

**Review by k0nserv — COMMENTED** [review]
Any idea why the snapshots don't seem to work correctly(No tint)?

File: Backpack/OverlayView/Classes/BPKOverlayView.m (hunk 5 above)
Comment (georgegillams) [reply Backpack/OverlayView/Classes/BPKOverlayView.m:116]: Changed to `1`. It seems that it's still needed sadly

File: Backpack/OverlayView/Classes/BPKOverlayView.m (hunk 6 above)
Comment (georgegillams) [reply Backpack/OverlayView/Classes/BPKOverlayView.m:138]: Fixed. AFAIK we should use these on all initialisers and setters, right?

File: Backpack/OverlayView/Classes/BPKOverlayView.m (hunk 7 above)
Comment (georgegillams) [reply Backpack/OverlayView/Classes/BPKOverlayView.m:31]: Refactor done

File: Backpack/OverlayView/Classes/BPKOverlayView.m (hunk 8 above)
Comment (georgegillams) [reply Backpack/OverlayView/Classes/BPKOverlayView.m:184]: Done

File: Backpack/OverlayView/Classes/BPKOverlayView.m (hunk 6 above)
Comment (k0nserv) [reply Backpack/OverlayView/Classes/BPKOverlayView.m:132]: Yeah all the ones that end up affect UI properties such as layer styles

File: Backpack/OverlayView/Classes/BPKBackgroundView.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+#import "BPKOverlayViewCornerStyles.h"
+#import "BPKOverlayViewOverlayTypes.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * `BPKBackgroundView` is a subclass of `UIView` which allows you to apply Skyscanner style tints to its content.
+ */
+IB_DESIGNABLE @interface BPKBackgroundView : UIView
```
[the remark this answers is not in the document]
Comment (georgegillams) [reply Backpack/OverlayView/Classes/BPKBackgroundView.h:29]: I was thinking about that. Is `BPKOverlayBackgroundView` better?

[the remark this answers is not in the document]
Comment (k0nserv) [reply Backpack/OverlayView/Classes/BPKBackgroundView.h:29]: Sounds good

**Review by k0nserv — APPROVED** [review]
LGTM

**Pull request merged**